Specific return types for values and values list

syastrov opened the pull request:
Resolves https://github.com/mkurnikov/django-stubs/issues/45

Whew... this was tougher to implement than I first imagined :)

syastrov committed:
Instead of using Literal types, overload QuerySet.values_list in the plugin. Fixes #43.

- Add a couple of extra type checks that Django makes:
  1) 'flat' and 'named' can't be used together.
  2) 'flat' is not valid when values_list is called with more than one field.

syastrov committed:
Determine better row types for values_list/values based on fields specified.

- In the case of values_list, we use a Row type with either a single primitive, Tuple, or NamedTuple.
- In the case of values, we use a TypedDict.
- In both cases, Any is used as a fallback for individual fields if those fields cannot be resolved.

A couple other fixes I made along the way:
- Don't create reverse relation for ForeignKeys with related_name='+'
- Don't skip creating other related managers in AddRelatedManagers if a dynamic value is encountered
  for related_name parameter, or if the type cannot be determined.

syastrov committed:
Merge branch 'master' into specific_return_types_for_values_and_values_list

syastrov committed:
Fix for TypedDict so that they are considered anonymous.

syastrov committed:
Clean up some comments.

syastrov committed:
Implement making TypedDict anonymous in a way that doesn't crash sometimes.

syastrov committed:
Fix flake8 errors.

syastrov committed:
Remove even uglier hack about making TypedDict anonymous.

mkurnikov commented on mypy_django_plugin/lookups.py:
```diff
@@ -0,0 +1,151 @@
+from mypy.types import Type, Instance
+
+from mypy_django_plugin import helpers
+from mypy_django_plugin.transformers.fields import get_private_descriptor_type
```
Maybe move `get_private_descriptor_type` to `helpers` as it's used in more than one file now?

mkurnikov commented on mypy_django_plugin/main.py:
```diff
@@ -156,7 +157,7 @@ def extract_and_return_primary_key_of_bound_related_field_parameter(ctx: Attribu
             return primary_key_type
 
-    is_nullable = helpers.get_fields_metadata(ctx.type.type).get(field_name, {}).get('null', False)
+    is_nullable = is_field_nullable(ctx.type.type, field_name)
```
Access method with `helpers.is_field_nullable` to prevent one import?

mkurnikov reviewed (changes requested):
Wow, this one is huge, thanks a lot! 

I'm not going to review it thoroughly, I'm going to trust your expertise and tests here. Just a few remarks. 

I'm going to run it against internal codebases tomorrow, and merge+release if everything is ok.

syastrov committed:
Address review comments. Write a few better comments inside tests.

syastrov committed:
Fix crash when running with mypyc ("interpreted classes cannot inherit from compiled") due to the way I extended TypedDictType.

- Implemented the hack in another way that works on mypyc.
- Added a couple extra tests of accessing 'id' / 'pk' via values_list.

syastrov committed:
Fix flake8 errors.

syastrov committed:
Merge branch 'master' into specific_return_types_for_values_and_values_list

syastrov committed:
Support annotation expressions (use type Any) for TypedDicts row types returned by values_list.

- Bonus points: handle values_list gracefully (use type Any) where Tuples are returned
  where some of the fields arguments are not string literals.
